src/os/bluestore/BlueFS.cc

```diff
@@ -3861,11 +3859,16 @@ int BlueFS::truncate(FileWriter *h, uint64_t offset)/*_WF_L*/
   if (offset > fnode.size) {
     ceph_abort_msg("truncate up not supported");
   }
-
+  unittest_inject_delay();
   _flush_bdev(h);
   {
     std::lock_guard ll(log.lock);
     std::lock_guard dl(dirty.lock);
+    std::lock_guard fl(h->file->lock);
+    if (h->file->deleted) {
```

Comment:
Shouldn't we protect other file->deleted checks with file lock. Namely in _flush_range_F() and _signal_dirty_to_log_D?

Reply:
Changed the logic.
All the cases we used to look at file->deleted should be covered now.

Reply:
Seems still missing for _signal_dirty_to_log_D()

Reply:
> Seems still missing for _signal_dirty_to_log_D()

This is ok, all action is under dirty lock, which is required for transition deleted=false => deleted=true.

The extra lock in code above `std::lock_guard fl(h->file->lock);` is because I felt like modifying fnode deserves having it.